Fix segfault for invalid axes in np.split

aseyboldt opened the pull request:
This adds a new function `normalize_axis` to `arrayobj.py`, and uses this function to verify axis arguments in `np.split`. There already is a similar function that could be wrapped in an intrinsic, but because the error message needs to be constant that seems more complicated than necessary to me.

I also use the new `normalize_axis` function to simplify `np.swapaxes` and `np.take_along_axis` a bit.

Fixes #8257

aseyboldt commented:
There are still some test failures, but maybe they are unrelated? The workflow fails even before the tests are started.

guilhermeleobas commented on numba/np/arrayobj.py:
```diff
@@ -266,6 +267,29 @@ def update_array_info(aryty, array):
                                           get_itemsize(context, aryty))
 
 
+def normalize_axis(func_name, ndim, axis):
+    """Constrain axis values to valid positive values."""
+    raise NotImplementedError()
+
+
+@overload(normalize_axis)
+def _normalize_axis_overloads(func_name, ndim, axis):
+    if not isinstance(func_name, Literal):
+        raise errors.TypingError("func_name must be a str literal.")
+
+    msg = f"{func_name.literal_value}: Axis out of bounds"
```
Instead of an overload, you can either use `@generated_jit` or `@register_jitable`

aseyboldt replied:
I switched to `generated_jit`. I don't see how I could use `register_jitable` though, because I think the error message needs to be known at compile time, and register_jitable can't dispatch on literal types, can it?

guilhermeleobas replied:
That's correct! But it should be possible after https://github.com/numba/numba/pull/8134 gets merged.

guilhermeleobas reviewed (commented):
HI @aseyboldt, thanks for working on this issue. I gave an initial review

aseyboldt commented on numba/np/arrayobj.py:
```diff
@@ -266,6 +267,29 @@ def update_array_info(aryty, array):
+
+@overload(normalize_axis)
+def _normalize_axis_overloads(func_name, ndim, axis):
+    if not isinstance(func_name, Literal):
```
fixed

aseyboldt commented on numba/np/arrayobj.py:
```diff
@@ -5641,8 +5662,10 @@ def impl(ary, indices_or_sections, axis=0):
+            axis = normalize_axis("np.split", ary.ndim, axis)
+            out = []
             prev = 0
+            cur = 0
```
fixed

aseyboldt commented:
Completely forgot about this PR :-)

> The reduced level of information in the error message is a little unfortunate - would you be able to modify the normalize_axis() function so it takes the function name and argument name and bakes them both into the exception message?

I changed the message to:
```
f"{func_name.literal_value}: Argument {arg_name.literal_value} out of bounds for dimensions of the array"
```

And reverted back to overload

gmarkall commented on numba/tests/test_np_functions.py:
```diff
@@ -2768,6 +2772,10 @@ def test_split_basic(self):
 
+        with self.assertRaises(ValueError) as raises:
+            njit(split)(np.ones(5), [3], axis=-3)
+        self.assertIn("np.split", str(raises.exception))
```
```suggestion
        self.assertIn("np.split: Argument axis out of bounds",
                      str(raises.exception))
```

gmarkall commented on numba/tests/test_np_functions.py:
```diff
@@ -4651,14 +4659,12 @@ def test_swapaxes_exception(self):
 
-        self.assertIn('The second argument "axis1" is out of bounds for array'
-                      ' of given dimension', str(raises.exception))
+        self.assertIn('np.swapaxes', str(raises.exception))
```
```suggestion
        self.assertIn('np.swapaxes: Argument axis1 out of bounds',
                      str(raises.exception))
```

gmarkall commented on numba/tests/test_np_functions.py:
```diff
@@ -4651,14 +4659,12 @@ def test_swapaxes_exception(self):
 
-        self.assertIn('The third argument "axis2" is out of bounds for array'
-                      ' of given dimension', str(raises.exception))
+        self.assertIn('np.swapaxes', str(raises.exception))
```
```suggestion
        self.assertIn('np.swapaxes: Argument axis2 out of bounds',
                      str(raises.exception))
```

gmarkall reviewed (changes requested):
Many thanks for the updates - I think this is looking pretty good. I have made some small suggestions on the diff just to ensure the reporting of the argument name as well as the function name - are you happy with these? If so, then once they are accepted I can approve this PR.

A quick note on process for the future - could you use merge commits and avoid rebasing / force-pushes please? GIthub reviews get a bit confused when there is a force push, which makes it harder to check what the changes since the last review were - many thanks in advance!

aseyboldt commented:
>  so, then once they are accepted I can approve this PR.

done

> A quick note on process for the future - could you use merge commits and avoid rebasing / force-pushes please?

Sure, I will to that in the future :-)

gmarkall reviewed (approved):
Many thanks for the fix!
